File: typescript/sskr/src/secret.ts
import { MIN_SECRET_LEN, MAX_SECRET_LEN } from './constants.js';
import { SskrError } from './error.js';

/**
 * A secret to be split into shares.
 *
 * Wraps a `Uint8Array` with length validation ensuring the data is between
 * {@link MIN_SECRET_LEN} and {@link MAX_SECRET_LEN} bytes and has even length.
 */
export class Secret {
    readonly #data: Uint8Array;

    private constructor(data: Uint8Array) {
        this.#data = data;
    }

    /**
     * Creates a new `Secret` from the given byte data.
     *
     * @param data - The secret data to be split into shares.
     * @returns A new `Secret` instance.
     * @throws {SskrError} If the length is less than {@link MIN_SECRET_LEN},
     *   greater than {@link MAX_SECRET_LEN}, or not even.
     */
    static create(data: Uint8Array): Secret {
        const len = data.length;
        if (len < MIN_SECRET_LEN) {
            throw SskrError.secretTooShort();
        }
        if (len > MAX_SECRET_LEN) {
            throw SskrError.secretTooLong();
        }
        if ((len & 1) !== 0) {
            throw SskrError.secretLengthNotEven();
        }
        return new Secret(Uint8Array.from(data));
    }

    /** The length of the secret in bytes. */
    get length(): number {
        return this.#data.length;
    }

    /** Whether the secret is empty. */
    get isEmpty(): boolean {
        return this.length === 0;
    }

    /** The raw secret data. */
    get data(): Uint8Array {
        return this.#data;
    }

    /** Returns true if this secret is equal to another. */
    equals(other: Secret): boolean {
        if (this.length !== other.length) return false;
        for (let i = 0; i < this.length; i++) {
            if (this.#data[i] !== other.#data[i]) return false;
        }
        return true;
    }
}

File: go/sskr/secret.go
package sskr

import "bytes"

// Secret is a secret to be split into shares.
type Secret struct {
	data []byte
}

// NewSecret creates a new Secret with the given data.
func NewSecret(data []byte) (Secret, error) {
	length := len(data)
	if length < MinSecretLen {
		return Secret{}, ErrSecretTooShort
	}
	if length > MaxSecretLen {
		return Secret{}, ErrSecretTooLong
	}
	if length&1 != 0 {
		return Secret{}, ErrSecretLengthNotEven
	}
	copied := make([]byte, len(data))
	copy(copied, data)
	return Secret{data: copied}, nil
}

// Len returns the length of the secret.
func (s Secret) Len() int {
	return len(s.data)
}

// IsEmpty reports whether the secret is empty.
func (s Secret) IsEmpty() bool {
	return s.Len() == 0
}

// Data returns a copy of the secret bytes.
func (s Secret) Data() []byte {
	copied := make([]byte, len(s.data))
	copy(copied, s.data)
	return copied
}

// bytes returns the underlying byte slice without copying.
// Internal use only; callers must not modify the returned slice.
func (s Secret) bytes() []byte {
	return s.data
}

// Clone returns a deep copy of the secret.
func (s Secret) Clone() Secret {
	copied := make([]byte, len(s.data))
	copy(copied, s.data)
	return Secret{data: copied}
}

// Equal reports whether two secrets contain identical bytes.
func (s Secret) Equal(other Secret) bool {
	return bytes.Equal(s.data, other.data)
}
